### PR TITLE
Fix some MergeFlags issues

### LINKED ISSUE #2961: Fix coercion of Strings to Lists in MergeFlags
This issue was originally created at: 2014-07-13 05:15:21.
This issue was reported by: `dirkbaechle`.
dirkbaechle said at 2014-07-13 05:15:21
>As reported on the mailing list: https://scons.tigris.org/ds/viewMessage.do?dsForumId=1272&viewType=browseAll&dsMessageId=2876001 the `MergeFlags()` method doesn't handle appending flags to already existing values very well. The coercion of types in `ParseFlags()` (as copied from the `Append` method) should get amended to properly handle the following use case:
```py
env.MergeFlags({'CCFLAGS' : '-whatever -O3'})
```
where `CCFLAGS` is initially empty.

dirkbaechle said at 2014-07-13 05:33:10
>added dependency

dirkbaechle said at 2014-07-13 07:46:05
>*** Issue 2402 has been marked as a duplicate of this issue. ***

dirkbaechle said this issue depends on #2962 at 2014-07-13 05:33:10.

dirkbaechle said this issue is duplicated by #2402 at 2014-07-13 07:46:05.

### LINKED ISSUE #3017: Mis-processing of GCC flags "--param arg1=val1 --param arg2=val2"
This issue was originally created at: 2015-12-24 15:16:27.
This issue was reported by: `hartwork`.
hartwork said at 2015-12-24 15:16:27
>Hi!
> 
>When building libmypaint [1] I ran into a bug with valid gcc CFLAGS
> 
>  CFLAGS="--param l1-cache-size=32 --param l2-cache-size=6144"
> 
>that make otherwise happy compilation fail.  I wonder if we are dealing with a
>bug in the way that scons processes CFLAGS or with libmypaint's specific use of
>scons.  I am hoping for your help with answering that question.
> 
>libmypaint's build system is accessing CFLAGS like this:
> 
>  # fgrep -Rw CFLAGS .
>  ./SConstruct:if os.environ.has_key('CFLAGS'):
>  ./SConstruct:   env['CCFLAGS'] += SCons.Util.CLVar(os.environ['CFLAGS'])
>  ./SConstruct:env.Append(CFLAGS='-std=c99')
>  ./SConstruct:env.Append(CFLAGS='-D_POSIX_C_SOURCE=200809L')
>  ./SConscript:    env.Append(CFLAGS='-fopenmp')
> 
>To reproduce the error I get (ruther down) you could run the following:
> 
>  # git clone https://github.com/mypaint/libmypaint.git
> 
>  # cd libmypaint
> 
>  # git checkout -b mybranch 379dbba7897815d9bad8bb068560376f03dcd12b
> 
>  # CFLAGS="--param l1-cache-size=32 --param l2-cache-size=6144" scons
>enable_gegl="true"
>  scons: Reading SConscript files ...
>  scons: done reading SConscript files.
>  scons: Building targets ...
>  Creating 'mypaint-config.h'
>  gcc -o brushmodes.os -c -std=c99 -D_POSIX_C_SOURCE=200809L -fopenmp -O3
>--param l1-cache-size=32 --param l2-cache-size=6144 -Wall -Wstrict-prototypes
>-Werror -g -fPIC -DHAVE_JSON_C -DHAVE_GETTEXT -I. -I/usr/include/json-c brushmodes.c
>  generate_cheaders(["mypaint-brush-settings-gen.h", "brushsettings-gen.h"],
>["generate.py", "brushsettings.py", "brushsettings.json"])
>  python2.7 generate.py mypaint-brush-settings-gen.h brushsettings-gen.h
>  Writing mypaint-brush-settings-gen.h
>  Writing brushsettings-gen.h
>  gcc -o fifo.os -c -std=c99 -D_POSIX_C_SOURCE=200809L -fopenmp -O3 --param
>l1-cache-size=32 --param l2-cache-size=6144 -Wall -Wstrict-prototypes -Werror -g
>-fPIC -DHAVE_JSON_C -DHAVE_GETTEXT -I. -I/usr/include/json-c fifo.c
>  gcc -o gegl/mypaint-gegl-surface.os -c -std=c99 -D_POSIX_C_SOURCE=200809L
>-fopenmp -O3 l1-cache-size=32 --param l2-cache-size=6144 -Wall
>-Wstrict-prototypes -Werror -g -pthread -fPIC -DHAVE_JSON_C -DHAVE_GETTEXT -I.
>-I/usr/include/json-c -I/usr/include/gegl-0.3 -I/usr/include/json-glib-1.0
>-I/usr/include/gio-unix-2.0 -I/usr/include/glib-2.0
>-I/usr/lib64/glib-2.0/include -I/usr/include/babl-0.1 gegl/mypaint-gegl-surface.c
>  gcc: error: l1-cache-size=32: No such file or directory
>  scons: *** [gegl/mypaint-gegl-surface.os] Error 1
>  scons: building terminated because of errors.
> 
>I am looking forward on your feedback.
> 
>Thanks and best, Sebastian
> 
>[1] https://github.com/mypaint/libmypaint
>[2] https://github.com/mypaint/libmypaint/issues/30

bdbaddog said at 2015-12-24 19:09:42
>Generally we ask users to bring their issues to the users mailing list to
>discuss the issue before filing a bug.
>You can see our bug policy here:
>http://scons.org/bugs.php
> 
>This will get far more eyes than filing the bug.
> 
>Can you try escaping the space between param and it's argument?

hartwork said at 2015-12-25 05:31:53
>When escaped I get this:
> 
># CFLAGS="--param\ l1-cache-size=32 --param\ l2-cache-size=6144" scons
>enable_gegl="true"                                                             
>                                                                
>scons: Reading SConscript files ...                                            
>                                                                               
>                                                       
>scons: done reading SConscript files.                                          
>                                                                               
>                                                       
>scons: Building targets ...
>gcc -o brushmodes.os -c -std=c99 -D_POSIX_C_SOURCE=200809L -fopenmp -O3 --param\
>l1-cache-size=32 --param\ l2-cache-size=6144 -Wall -Wstrict-prototypes -Werror
>-g -fPIC -DHAVE_JSON_C -DHAVE_GETTEXT -I. -I/usr/include/json-c brushmodes.c
>gcc: error: unrecognized command line option ‘--param l1-cache-size=32’
>gcc: error: unrecognized command line option ‘--param l2-cache-size=6144’
>scons: *** [brushmodes.os] Error 1
>scons: building terminated because of errors.
> 
> 
>Same with double back-slash.
> 
>(Since in some distros CFLAGS are configured system-wide, making the escaped
>version work would not be a real fix since it would break other build systems then.)

### PR DESCRIPTION
**Pull request by mwichmann**
- signature/return did not match documentation or existing usage.  Removed `dict` arg; method now does not return `self`.
- merging `--param` arguments did not work (`ParseFlags` issue). partly fixes #3017
- passing a dict to merge where the values are strings failed. fixes #2961

Tests are beefed up to confirm these.

No direct doc impacts from these changes, but some docstrings updated.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

### CODE REVIEW

File: SCons/Environment.py [hunk 1]
```diff
@@ -1623,7 +1644,7 @@ def ParseConfig(self, command, function=None, unique=1):
         """
         if function is None:
             def parse_conf(env, cmd, unique=unique):
-                return env.MergeFlags(cmd, unique)
+                env.MergeFlags(cmd, unique)
```
Comment (bdbaddog) [inline SCons/Environment.py]: Why remove the return?

File: SCons/Environment.py [hunk 2]
```diff
@@ -819,24 +835,30 @@ def append_define(name, dict = dict):
...
         return dict
 
-    def MergeFlags(self, args, unique=1, dict=None):
-        """
-        Merge the dict in args into the construction variables of this
-        env, or the passed-in dict.  If args is not a dict, it is
-        converted into a dict using ParseFlags.  If unique is not set,
-        the flags are appended rather than merged.
-        """
+    def MergeFlags(self, args, unique=True):
```
Comment (bdbaddog) [inline SCons/Environment.py:842]: Isn't self really env?

Comment (mwichmann) [reply SCons/Environment.py:842]: Yes - an instance of SubstitutionEnvironment or one of its children.

File: SCons/Environment.py (hunk 1 above)
Comment (mwichmann) [reply SCons/Environment.py]: it doesn't really have anything to return, but there's no harm in keeping the return statement either - MergeFlags does not return anything specific, it modifies env in place.

**Comment by mwichmann** [comment]
I propose *not* address the Sider issues, that's just what the tests do, a "star" import.

File: CHANGES.txt [hunk 3]
```diff
@@ -35,6 +35,10 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
       OverrideEnvironment. Enable env.setdefault() method, add tests.
     - Raise an error if an option (not otherwise consumed) is used which
       looks like an abbreviation of one one added by AddOption. (#3653)
+    - Fix three issues in MergeFlags: signature/return did not match
+      documentation or existing usage; merging --param arguments did
+      not work (issue #3107); passing a dict to merge where the values
+      are strings failed (issue #2961).
```
Comment (bdbaddog) [inline CHANGES.txt]: Add a note that it no longer returns the passed environment. Then should be good to merge.

**Pull request merged**